### PR TITLE
Mesh-based cutter class with growth capability

### PR DESCRIPTION
**Pull request by liuusu**
A new mesh-based cutter class is developed for the XFEM module. Two test cases are added: (1) edge_crack_3d_mesh.i is a slightly changed version of edge_crack_3d.i, in which the initial crack is defined by a mesh of triangular elements; and (2) mesh_grow.i is a new test example, in which the mesh cutter is allowed to grow following a predefined growth function.
Ref #9584

### CODE REVIEW

**Comment by liuusu** [comment]
@bwspenc   Shows "Developer needed to activate" again.  Will come back tomorrow to see if all tests pass.  I can revise the code based on what you are doing on the cutters, before merging.

**Comment by permcody** [comment]
I looked through this briefly and don't see any blatant problems. I would say from a general perspective that there is a lot more opportunity to clean up all of the geometry related operations. There are a lot of geometry routines in libMesh and MOOSE already and some of the methods in XFEMFuncs.C could probably be refactored to use more of those methods. Also, having richer overloads would help so you don't have to do things like convert points to vectors before passing them into geometry processing methods.

However, you are building off from what's already there so it's really up to the module maintainers to decide longer time development and cleanup strategies. I'll leave this to @bwspenc for the actual review and merge.

Thanks

File: modules/xfem/src/userobjects/MeshCut3DUserObject.C
Comment (bwspenc) [inline modules/xfem/src/userobjects/MeshCut3DUserObject.C]: `_boundary_map` would be better as a container for the CutPoint objects (`std::map<CutPoint>`) rather than a container of shared pointers. Also, I think you should use `this_node->id()` as the key in the map, rather than `i`.

File: modules/xfem/include/userobjects/GeometricCutUserObject.h
Comment (bwspenc) [inline modules/xfem/include/userobjects/GeometricCutUserObject.h]: This `CutPoint` struct seems really specific to your mesh-based cutter, so it would be better to put it in the header of that class.

File: modules/xfem/src/userobjects/MeshCut3DUserObject.C
Comment (bwspenc) [inline modules/xfem/src/userobjects/MeshCut3DUserObject.C]: Be consistent in use of `int` and `unsigned int` here.

File: modules/xfem/src/userobjects/MeshCut3DUserObject.C
Comment (bwspenc) [inline modules/xfem/src/userobjects/MeshCut3DUserObject.C]: I know that this is just a placeholder for doing something more general, but I hate having a hardcoded growth function here. You could supply references to three `Function` objects, that can supply the three components of the growth direction as a function of position and time. In the input file, those can just be parsed functions to do what you are doing here. That may not be the way we want to do this in the end, but at least it would be definable in the input file.

**Review by bwspenc — CHANGES_REQUESTED** [review]
Sorry for not getting to this review sooner.  This is shaping up nicely, and I'm really excited about this capability. I don't think any of my issues will be too hard to sort out.

File: modules/xfem/include/userobjects/GeometricCutUserObject.h
Comment (liuusu) [reply modules/xfem/include/userobjects/GeometricCutUserObject.h]: You are right.  I had some problems and added this as a fix.  Now it's not needed.  Thanks.

File: modules/xfem/src/userobjects/MeshCut3DUserObject.C
Comment (liuusu) [reply modules/xfem/src/userobjects/MeshCut3DUserObject.C]: You are right.  This is the number of steps to grow the mesh.  However, assuming XFEM and this code are fully integrated, this number of steps is exactly the same as the number of steps to change the cuts.  The input mesh serves as the initial cut.  The cut (which is also the mesh) will be updated based on stress intensity factors.  Do you still want me to remove this, and hard code a number in the C file?

File: modules/xfem/src/userobjects/MeshCut3DUserObject.C
Comment (liuusu) [reply modules/xfem/src/userobjects/MeshCut3DUserObject.C]: What we want is: (1) structural mesh is 3D. (2) Cutter mesh is 2D. (3) Cutter Mesh has Tri only.  In this revision, I added the checker for (1) and (2).  I had (3) before.  Let's me know if this isn't what you want.

File: modules/xfem/src/userobjects/MeshCut3DUserObject.C
Comment (liuusu) [reply modules/xfem/src/userobjects/MeshCut3DUserObject.C]: I got rid of the definition of CutPoint, and simplified the code here.  It is better now.

File: modules/xfem/src/userobjects/MeshCut3DUserObject.C
Comment (liuusu) [reply modules/xfem/src/userobjects/MeshCut3DUserObject.C]: I had to overload the < operator for CutEdge to do std::set<CutEdge>.  To not modify GeometricCutUserObject.h, I put the < overload in MeshCut3DUserObject.h.

Comment (bwspenc) [reply modules/xfem/src/userobjects/MeshCut3DUserObject.C]: I think you would want to put that in CutEdge.  I guess I'd have to see what you changed.

File: modules/xfem/src/userobjects/MeshCut3DUserObject.C
Comment (bwspenc) [reply modules/xfem/src/userobjects/MeshCut3DUserObject.C]: That sounds right.

File: modules/xfem/src/userobjects/MeshCut3DUserObject.C
Comment (bwspenc) [reply modules/xfem/src/userobjects/MeshCut3DUserObject.C]: Since you're using this for testing, I'd say leave it in there.  We can remove it later on when we don't need it.  Since you have the capability to grow the mesh in your cutter already, we definitely want to test that.

**Comment by bwspenc** [comment]
@liuusu I'm going through your code now. It looks like the test you just added is failing in debug (the "Modules debug" target).  I wouldn't worry about the "Test Intel" failure -- that seems unrelated to your commit, and I'll force those tests to re-run to see if that has been fixed.

**Comment by bwspenc** [comment]
@liuusu I just tried building your code and running the test in debug, and that test passes for me on my Mac. I'll try it on Linux tomorrow to see if I can figure out why it's breaking.

**Comment by liuusu** [comment]
Thanks for your help. I did everything on mac.  It seems some unrelated 
tests failed.  If I can do anything, please let me know.

Ben Spencer wrote, On 2/8/2018 6:07 PM:

**Comment by liuusu** [comment]
@bwspenc  Thanks for your help. I did everything on mac.  It seems some unrelated tests failed?  If I can do anything, please let me know.

**Comment by permcody** [comment]
The Intel failure was unrelated. Please ignore.

**Comment by bwspenc** [comment]
I've managed to replicate the error on Linux, but I'm struggling to get a backtrace or something to figure out where it's coming from. It doesn't give any valgrind errors.  I can run it in the debugger (gdb), but I'm not coming up with much. If anyone has any suggestions, I'd welcome them. This is the error:

/ssd/opt/moose/gcc-7.2.0/include/c++/7.2.0/debug/vector:417:
Error: attempt to subscript container with out-of-bounds index 0, but
container only holds 0 elements.

Objects involved in the operation:
    sequence "this" @ 0x0x7fffffffd450 {
      type = std::__debug::vector<libMesh::Point, std::allocator<libMesh::Point> >;
    }

Program received signal SIGABRT, Aborted.
0x00007fffea31b428 in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:54
54	../sysdeps/unix/sysv/linux/raise.c: No such file or directory.

**Comment by permcody** [comment]
Did you break on "MPI_Abort"? That's usually the best way to catch the
error before it hits the exception handler and all of the useful stack
information gets discarded. You should be able to pin this down with the
right breakpoint.

Cody

**Comment by bwspenc** [comment]
@permcody Thanks for the hint! That did the trick, and I know where it's failing now.

File: modules/xfem/src/userobjects/MeshCut3DUserObject.C
Comment (bwspenc) [inline modules/xfem/src/userobjects/MeshCut3DUserObject.C]: Remove the `vertices.clear();`  That's what was making that test fail in debug. It zeros out the size of the vector, so when you access an element of it in the loop below, you're out of the bounds of the vector.

File: modules/xfem/src/userobjects/MeshCut3DUserObject.C
Comment (bwspenc) [inline modules/xfem/src/userobjects/MeshCut3DUserObject.C]: Remove `node_id.clear()` and `is_node_on_boundary.clear()`.  Same deal as the other vector you were clearing.

**Review by bwspenc — CHANGES_REQUESTED** [review]
I got a chance to get back this and find where that test failure came from.  See my comments for the specific lines of code to change. They're really easy fixes.  Once you fix those, I think this will be good to go.

**Comment by liuusu** [comment]
Thanks for helping me find the bugs.  I have made the changes and pushed the code.

File: modules/xfem/src/userobjects/MeshCut3DUserObject.C
Comment (permcody) [inline modules/xfem/src/userobjects/MeshCut3DUserObject.C]: const Point & p

File: modules/xfem/include/userobjects/MeshCut3DUserObject.h
Comment (permcody) [inline modules/xfem/include/userobjects/MeshCut3DUserObject.h]: I feel like a lot of these geometry methods are re-inventing the wheel. Before we add another round of geometry code to the modules, it would be nice to review what might be overlapping. Take a look in libMesh at the `Plane`, `Line` and related objects. We can already do normals, projections, with basic types. It's in libMesh. On top of that, we've implemented additional search routines in our geometric search system. Projections of nodes onto planes, lines onto planes, intersections, etc. Now, I'm see it all again for this object. We have to get better about code reuse. I'm more than willing to help find this or assist you in finding these so we don't keep adding to the problem.

File: modules/xfem/src/base/XFEMFuncs.C
Comment (permcody) [inline modules/xfem/src/base/XFEMFuncs.C]: PI is defined in libMesh. It's also defined in some of these other stand alone XFEM utility files.

File: modules/xfem/src/userobjects/MeshCut3DUserObject.C
Comment (permcody) [inline modules/xfem/src/userobjects/MeshCut3DUserObject.C]: Not safe for DistributedMesh

File: modules/xfem/src/userobjects/MeshCut3DUserObject.C
Comment (permcody) [inline modules/xfem/src/userobjects/MeshCut3DUserObject.C]: We can already fine normals to planes. It's in libMesh.

File: modules/xfem/include/userobjects/MeshCut3DUserObject.h
Comment (permcody) [inline modules/xfem/include/userobjects/MeshCut3DUserObject.h]: With as much capability as we are adding, eventually it would be nice to migrate this capability to the ElementalUserObject system. There's a lot of time and complexity going into these objects to be put under a GeneralUserObject. None of this code will scale.

File: modules/xfem/src/userobjects/MeshCut3DUserObject.C
Comment (permcody) [inline modules/xfem/src/userobjects/MeshCut3DUserObject.C]: No DistributedMesh friendly.

File: modules/xfem/src/userobjects/MeshCut3DUserObject.C
Comment (permcody) [inline modules/xfem/src/userobjects/MeshCut3DUserObject.C]: No need to store into a Point. Nodes can do norms too. Also, the reason you'd get a nullptr is because of DistributedMesh. I like the asserts!

**Review by permcody — COMMENTED** [review]
@bwspenc - I'm going to leave all this code up to you, but I feel like we are taking on a lot of duplicate and restrictive code in XFEM. None of it is scalable or can run with Distributed Mesh. We need to talk about longer term plans for all of this capability.

**Comment by permcody** [comment]
Tag @friedmud

File: modules/xfem/include/userobjects/MeshCut3DUserObject.h
Comment (bwspenc) [reply modules/xfem/include/userobjects/MeshCut3DUserObject.h]: I have a PR (#10557) that moves all of these classes over to be based on ElementUserObject. That PR breaks a test, and it has been sitting there for a while because we were trying to sort out the bug that was causing it. We almost have a fix for that, so this will happen soon.  This PR and that one will have a conflict, so one of them needs to go in first. I'd like to get this in first, and then get it to play nice with the move to ElementUserObject, which won't be hard.

**Comment by bwspenc** [comment]
@permcody The reasons why XFEM doesn't work with distributed mesh are a lot bigger than this class. I wasn't aware of the `Mesh::errorIfDistributedMesh()` function, but I'd say we should just put that call in XFEM.C to cover everything. The fundamental problem there is that we run the cutting algorithm on a serialized version of the mesh on every processor.

I think the concerns about this part of the code being scalable should go away once I convert this over to be based on ElementUserObject as part of my other PR in progress.  That PR is a step toward making this all more DistributedMesh friendly. I'd say we shouldn't worry about that right now.

File: modules/xfem/src/base/XFEMFuncs.C
Comment (bwspenc) [reply modules/xfem/src/base/XFEMFuncs.C]: This obviously here because this code was straight-up copied without modification from John Burkhardt. I'm totally on-board with using geometric routines in libmesh if they are available instead of these. Since this only adds one function to a large set of functions like this, my opinion is that it would be better to tackle replacing all of these functions as a separate task rather than figure out whether we have the needed functionality in libMesh and use that as part of this PR.  This PR has been in the review phase for a long time, and I'd like to get this code in now and have our team start working on cleaning a few things like that up rather than asking more of @liuusu here.

File: modules/xfem/src/userobjects/MeshCut3DUserObject.C
Comment (bwspenc) [reply modules/xfem/src/userobjects/MeshCut3DUserObject.C]: It's important to note here that `_cut_mesh` is not the main FE mesh.  It's a topologically 2D mesh used to define cutting planes for a 3D FE mesh. It will always be significantly smaller than the main solution mesh, and I doubt that it would ever make sense for it to be distributed.

**Comment by bwspenc** [comment]
@permcody To avoid taking any more of @liuusu's time on this, @jiangwen84 is going to take a little time today to address the issues you pointed out.  He'll either do a PR onto this branch with his changes, or we'll close out this PR and he'll do a separate PR that has this code with his changes on top of it.

**Comment by permcody** [comment]
Fair enough, I won't hold this up.

File: modules/xfem/include/userobjects/GeometricCutUserObject.h
Comment (liuusu) [reply modules/xfem/include/userobjects/GeometricCutUserObject.h]: I hope I understood your comment correctly.  With the following changes, now (1,3)<(2,2) and (1,10)<(2,3).

I changed from: 
  return l1 < r1 || l2 < r2;

to:
  if (l1 != r1)
    return l1 < r1;
  else
    return l2 < r2;

File: modules/xfem/src/userobjects/MeshCut3DUserObject.C
Comment (liuusu) [reply modules/xfem/src/userobjects/MeshCut3DUserObject.C]: Applied range-based loops to elem and node loops where appropriate.

File: modules/xfem/src/userobjects/MeshCut3DUserObject.C
Comment (liuusu) [reply modules/xfem/src/userobjects/MeshCut3DUserObject.C]: Did this one.

File: modules/xfem/src/userobjects/MeshCut3DUserObject.C
Comment (liuusu) [reply modules/xfem/src/userobjects/MeshCut3DUserObject.C]: Changed to std::array.

File: modules/xfem/src/userobjects/MeshCut3DUserObject.C
Comment (liuusu) [reply modules/xfem/src/userobjects/MeshCut3DUserObject.C]: Changed to std::array.

File: modules/xfem/src/userobjects/MeshCut3DUserObject.C
Comment (liuusu) [reply modules/xfem/src/userobjects/MeshCut3DUserObject.C]: Did this for center and normal.

File: modules/xfem/src/userobjects/MeshCut3DUserObject.C
Comment (liuusu) [reply modules/xfem/src/userobjects/MeshCut3DUserObject.C]: Done

File: modules/xfem/include/userobjects/MeshCut3DUserObject.h
Comment (liuusu) [reply modules/xfem/include/userobjects/MeshCut3DUserObject.h]: Thanks for offering the help.  I found "Plane" and "ElemCutter" in libMesh quite useful, and I even started making a couple of changes.  Then I realized that all methods of making cuts and finding normals in this PR (MeshCut3DUserObject) were coded based on the existing code in GeometricCut3DUserObject. It seems to me that updating these geometric methods in both objects with the libMesh objects might be a separate Issue and PR.   Other than this, I have done changes to address the comments. Please advise if I should push what I have (and I would be glad to help with updating geometric methods later), or continue changes with the libMesh objects. Either is fine to me. @permcody @bwspenc @jiangwen84

File: modules/xfem/include/userobjects/GeometricCutUserObject.h
Comment (permcody) [reply modules/xfem/include/userobjects/GeometricCutUserObject.h]: > if (l1 != r1)
return l1 < r1;
else
return l2 < r2;

Yes, that makes sense, except you need to get rid of all the swap logic. You only need the comparison logic.

Comment (liuusu) [reply modules/xfem/include/userobjects/GeometricCutUserObject.h]: @permcody The swap logic is to make an edge with two nodes unique in comparison.  For example, (1,3) and (3,1) are geometrically the same edge.  We probably don't want the overloaded operator to give (1,3)<(2,4) true and (3,1)<(2,4) false. I used the swap logic to avoid these.  Does this make sense?

Comment (permcody) [reply modules/xfem/include/userobjects/GeometricCutUserObject.h]: That makes sense but it terms of storing items in a map I'm not sure it really matters.

I'm more concerned about confusing the map. If during the comparison for inserting an item you tell the map to go into the left tree for one case and into the right tree for another sometime you can insert items in the wrong place making the lookup not work like you might expect. That might not happen here, I'd have to sit down and really work through a few cases, but I'm concerned you might fall into one of those cases. If you think it's correct, then by all means, leave it.

Comment (liuusu) [reply modules/xfem/include/userobjects/GeometricCutUserObject.h]: What you wrote in the email makes sense to me.  I have removed the swap logic from this operator, and added it to my code where all CutEdges are stored.  This will make this operator more universal (not potentially confusing the map as you described in the email).  And it meets my needs too (avoid storing (1,3) and (3,1) for the same edge). Below is the updated code. Thanks!

inline bool
operator<(const CutEdge & lhs, const CutEdge & rhs)
{
  if (lhs.id1 != rhs.id1)
    return lhs.id1 < rhs.id1;
  else
    return lhs.id2 < rhs.id2;
}

File: modules/xfem/include/userobjects/MeshCut3DUserObject.h
Comment (liuusu) [reply modules/xfem/include/userobjects/MeshCut3DUserObject.h]: I have used "Plane.unit_normal" to calculate the normal to an element. This removes a method "findElemNormal", improving this PR. For other methods relating to cutting, I will not change in this PR because of the reason stated above. I will push what I have for your review. @permcody @bwspenc @jiangwen84

File: modules/xfem/src/userobjects/MeshCut3DUserObject.C
Comment (liuusu) [reply modules/xfem/src/userobjects/MeshCut3DUserObject.C]: Made the change and removed this method.

File: modules/xfem/src/userobjects/MeshCut3DUserObject.C
Comment (liuusu) [reply modules/xfem/src/userobjects/MeshCut3DUserObject.C]: Done.

**Review by bwspenc — APPROVED** [review]

**Comment by bwspenc** [comment]
Sorry for taking so long to get to this.  I think you've addressed all of the issues raised, so I'm going to merge this. Thank you for working with us on resolving those issues.  I'm excited to have this capability available!

The only outstanding issues I see are that @permcody pointed out a few operations on `_cut_mesh` not being safe for `DistributedMesh`, but it is created as a secondary `ReplicatedMesh` within this class, and I don't think it would ever make sense for it to be distributed.

In a subsequent commit, I'll add a general check on XFEM to make sure that the solution mesh isn't distributed. We'll also definitely look into doing a general replacement of the geometry operations in XFEM with available functions in libMesh.

**Pull request merged**
